File: 98.validate-binary-search-tree.py
#
# @lc app=leetcode id=98 lang=python3
#
# [98] Validate Binary Search Tree
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        # three conditions need to be true
        # root.val is within upperbound and lowerbound
        # left tree is valid
        # right tree is valid
        return self.isValidRoot(root, float("-inf"), float("inf"))

    def isValidRoot(self, root, lowerbound, upperbound):
        if root is None:
            return True
        return (
            self.isValidRoot(root.left, lowerbound, root.val)
            and self.isValidRoot(root.right, root.val, upperbound)
            and (lowerbound < root.val < upperbound)
        )


# @lc code=end

File: 98.validate-binary-search-tree.go
/*
 * @lc app=leetcode id=98 lang=golang
 *
 * [98] Validate Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
package main

import "math"

/*
define a recursive function validate(root *TreeNode, lower, upper int) bool {
 1. check if root in the bound
 2. left subtree should return true (validated)
 3. right subtree should return true
    Return 1 && 2 && 3
    }
*/
func isValidBST(root *TreeNode) bool {
	var validate func(root *TreeNode, lower, upper int) bool
	validate = func(root *TreeNode, lower, upper int) bool {
		if root == nil {
			return true
		}
		if root.Val <= lower || root.Val >= upper {
			return false
		}
		if !validate(root.Left, lower, root.Val) {
			return false
		}
		if !validate(root.Right, root.Val, upper) {
			return false
		}
		return true
	}

	return validate(root, -math.MaxInt, math.MaxInt)
}

// @lc code=end
